### PR TITLE
Added Google Gemini as a custom LLM provider

### PR DESCRIPTION
**Pull request by haseeb-heaven**
**Introducing Gemini Language Support!** 

This pull request is for **Gemini** support through google **Generative-AI** library for [Feature Request#1126](https://github.com/BerriAI/litellm/issues/1126)

**Key Enhancements:**

- Added `gemini.py` file mirroring `palm.py` for consistent code structure and maintainability.
- Updated `provider_list` in `init.py` to include `"gemini"` as a recognized provider.
- Implemented unit tests in `test_completion.py` and `test_streaming.py` for Gemini's completion and streaming functionality.
- Integrated Gemini models under `gemini/<model-name>` namespace in `model_prices_and_context_window.json` for clarity.

### CODE REVIEW

File: litellm/llms/gemini.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    ## CALCULATING USAGE - baseten charges on time, not tokens - have some mapping of cost here. 
+    prompt_tokens = len(
+        encoding.encode(prompt)
+    ) 
+    completion_tokens = len(
+        encoding.encode(model_response["choices"][0]["message"].get("content", ""))
+    )
+
+    model_response["created"] = int(time.time())
+    model_response["model"] = "palm/" + model
```
Comment (ishaan-jaff) [inline litellm/llms/gemini.py]: this should be `gemini/`

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
- We need to actually support `optional_params` see `get_optional_params()` in utils.py
- We need exception mapping for gemini, see `exception_type()` in utils.py

File: litellm/llms/gemini.py (hunk 1 above)
Comment (haseeb-heaven) [reply litellm/llms/gemini.py]: Right will make changes

**Comment by haseeb-heaven** [comment]
> * We need to actually support `optional_params` see `get_optional_params()` in utils.py
> * We need exception mapping for gemini, see `exception_type()` in utils.py

Is this specific for Gemini model? because i think if palm has those then Gemini will also support them

File: litellm/llms/gemini.py (hunk 1 above)
Comment (haseeb-heaven) [reply litellm/llms/gemini.py]: Done!

**Comment by haseeb-heaven** [comment]
> * We need to actually support `optional_params` see `get_optional_params()` in utils.py
> * We need exception mapping for gemini, see `exception_type()` in utils.py

I have added exception handling for Gemini but how to add optional_params for Gemini?
Seems like Gemini.py already has optional_params passed from main file

File: litellm/utils.py [hunk 2]
```diff
@@ -4531,6 +4531,36 @@ def exception_type(
                             llm_provider="palm",
                             response=original_exception.response
                         )
```
Comment (ishaan-jaff) [inline litellm/utils.py]: we also need a block for getting optional params for `gemini`

example here for palm
https://github.com/BerriAI/litellm/blob/707cd0364e10344d38ff621225a264e5ea987ea4/litellm/utils.py#L2463

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
requested 1 more change. otherwise lgtm!

**Comment by ishaan-jaff** [comment]
cc @krrishdholakia since you were working on the issue

File: litellm/utils.py (hunk 2 above)
Comment (haseeb-heaven) [reply litellm/utils.py]: Done added that now.

File: litellm/utils.py [hunk 3]
```diff
@@ -2431,11 +2431,17 @@ def _map_and_modify_arg(supported_params: dict, provider: str, model: str):
...
             optional_params["presencePenalty"] = {"scale": presence_penalty}
+    
     elif custom_llm_provider == "palm": # https://developers.generativeai.google/tutorials/curl_quickstart
         ## check if unsupported param passed in 
         supported_params = ["temperature", "top_p", "stream", "n", "stop", "max_tokens"]
         _check_valid_arg(supported_params=supported_params)
         
+    elif custom_llm_provider == "gemini": 
+        ## check if unsupported param passed in
+        supported_params = ["temperature", "topK", "topP", "maxOutputTokens", "stopSequences"]
```
Comment (ishaan-jaff) [inline litellm/utils.py]: this should be 
`supported_params = ["temperature", "top_p", "stream", "n", "stop", "max_tokens"]`

the `supported_params` are supported params to litellm.completion https://docs.litellm.ai/docs/completion/input

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
optional params needs a fix

File: litellm/utils.py (hunk 3 above)
Comment (haseeb-heaven) [reply litellm/utils.py]: But these are for Gemini API i checked from their Rest API response its same params from there.

**Comment by haseeb-heaven** [comment]
> optional params needs a fix

in which file can you point that in commit?

File: litellm/utils.py (hunk 3 above)
Comment (ishaan-jaff) [reply litellm/utils.py]: no but support_params needs to be the args sent to litellm https://docs.litellm.ai/docs/completion/input#input-params-1 

it needs to be one of these params here.

**Comment by ishaan-jaff** [comment]
> > optional params needs a fix
> 
> in which file can you point that in commit?

utils.py

**Pull request closed without merging**